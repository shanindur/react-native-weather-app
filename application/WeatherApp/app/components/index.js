export {default as SquareButton} from './button/squareButton';
export {default as RoundButton} from './button/roundButton';
export {default as ImageCapture} from './imageCapture';
export {default as AlertModalMultiple} from './alertModalMultiple';
export {default as Avatar} from './avatar';
export {default as TextInput} from './textInput';

