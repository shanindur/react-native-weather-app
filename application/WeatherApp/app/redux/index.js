export {default as Action} from './actions/action';
export {default as ActionType} from './actions/actionType';
export {default as Reducer} from './reducers/reducer';
export {default as Store} from './store/store';


