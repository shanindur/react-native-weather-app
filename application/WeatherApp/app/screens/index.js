export {default as SplashScreen} from './splash';
export {default as SignInScreen} from './auth/signin';
export {default as SignUpScreen} from './auth/signup';
export {default as MainScreen} from './main';
export {default as ProfileScreen} from './profile';
export {default as NoInternet} from './noInternet';
