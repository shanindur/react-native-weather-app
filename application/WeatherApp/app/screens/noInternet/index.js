/**
 * Weather App
 *
 * No Internet Screen
 *
 * @author       Shanindu Rajapaksha
 *
 * @version      0.1.0 2020-Jun-29
 */
import React from 'react';
import { SafeAreaView, ImageBackground, View, Text } from 'react-native';

import styles from './style';


const NoInternet = () => (
	<SafeAreaView>
		<ImageBackground style={styles.background} source={require('../../assets/images/splash.jpg')}>
			<View style={styles.container}>
				<Text style={styles.titleText}>No internet connection.</Text>
				<Text style={styles.bodyText}>Please restore your Internet connection</Text>
			</View>
		</ImageBackground>
	</SafeAreaView>
);

export default NoInternet;
