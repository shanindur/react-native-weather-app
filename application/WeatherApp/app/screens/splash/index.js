/**
 * WeatherApp
 *
 * Splash Screen
 *
 * @author       Shanindu Rajapaksha
 *
 * @version      0.1.0 2020-Jun-29
 */

import React from 'react';
import {SafeAreaView, ImageBackground } from 'react-native';
import styles from './styles';

const App = () => (
	<SafeAreaView>
		<ImageBackground style={styles.background} source={require('../../assets/images/splash.jpg')}/>
	</SafeAreaView>
);


export default App;
