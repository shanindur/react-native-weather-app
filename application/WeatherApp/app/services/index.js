export {default as StorageService} from './storageService';
