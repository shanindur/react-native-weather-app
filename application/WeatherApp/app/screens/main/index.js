/**
 * Weather App
 *
 * Main
 *
 * @author       Shanindu Rajapaksha
 *
 * @version      0.1.0 2020-Jul-01
 */
import React, { useEffect, useState, useRef } from 'react';
import {
	SafeAreaView,
	ScrollView,
	TouchableOpacity,
	View,
	Text,
	Image,
	Dimensions,
	RefreshControl,
	BackHandler,
	Alert
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import {useIsFocused} from '@react-navigation/native';
import styles from './styles';
import { Constants, Colors } from '../../util';
import {Avatar} from '../../components';
import {StorageService} from '../../services';
import axios from 'axios';
import {
	BarChart
} from "react-native-chart-kit";
import Geolocation from '@react-native-community/geolocation';



const Main = props => {
	const dispatch = useDispatch();
	const {isSignedIn} = useSelector(state => state);

	const isFocused = useIsFocused();

	const [ fetchingData, setFetchingData ] = useState(true);
	const [ currentData, setCurrentData ] = useState(null);
	const [ monthlyLabel, setMonthlyLabel ] = useState([]);
	const [ monthlyData, setMonthlyData ] = useState([]);
	const [ proImage, setProImage ] = useState();
	const [ refreshing, setRefreshing ] = useState(false);

	const fetchCurrentData = (lat, lon) => {
		setFetchingData(true);
		axios.get(Constants.CURRENT_URL, {
			params: {
				lat: lat,
				lon: lon,
				appid: Constants.API_KEY[0]
			}
		})
			.then(response => {
				console.log(response.data);
				setCurrentData(response.data);
			})
			.catch(error => {
				console.log(error);
			})
			.finally(() => {
				setFetchingData(false);
			});
	};

	const fetchMonthData = (lat, lon) => {
		setFetchingData(true);
		axios.get(Constants.MONTHLY_URL, {
			params: {
				lat: lat,
				lon: lon,
				appid: Constants.API_KEY[1]
			}
		})
			.then(response => {
				const label = response.data.list.map(elem => {
					const d = new Date(elem.dt * 1000);
					return d.getDate() + '/' + d.getMonth();
				});
				const data = response.data.list.map(elem => (elem.temp.average - 273.15).toFixed(0));
				setMonthlyLabel(label);
				setMonthlyData(data);
				console.log('monthData', label, data);
			})
			.catch(error => {
				console.log(error);
			})
			.finally(() => {
				setFetchingData(false);
			});
	};

	const dataChart = {
		labels: monthlyLabel,
		datasets: [
			{
				data: monthlyData
			}
		]
	};

	const chartConfig = {
		backgroundColor: Colors.primaryColour,
		backgroundGradientFrom: Colors.graphGradientFrom,
		backgroundGradientTo: Colors.primaryColour,
		decimalPlaces: 0, // optional, defaults to 2dp
		color: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`,
		labelColor: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`,
		style: {
			borderRadius: 16
		},
		propsForDots: {
			r: "6",
			strokeWidth: "2",
			stroke: Colors.white
		}
	};

	const getLocation = () => {
		Geolocation.getCurrentPosition((position) => {
			const { coords: { latitude, longitude } } = position;

			fetchCurrentData(latitude, longitude);
			fetchMonthData(latitude, longitude);

		}, (err) => {
			console.log(err);
			Alert.alert(
				'',
				'We couldn\'t fetch your location',
				[{
					text: 'Cancel',
					style: 'cancel',
					onPress: () => BackHandler.exitApp()
				}, {
					text: 'Try again',
					onPress: () => getLocation()
				}]
			);
			return;
		}, {
			timeout: 60000,
			maximumAge: 3600000, // Frequent Location capture timeouts with this
			enableHighAccuracy: false // `false` times out in simulators
		});
	};

	const getUserData = async() => {
		const user = await StorageService.getData('USER');
		if (user){
			setProImage(user.profile);
		}

	};

	const onRefresh = () => {
		setRefreshing(true);
		getLocation();
		getUserData();
		setRefreshing(false);
	};

	useEffect(() => {
		if (!isSignedIn){
			props.navigation.replace('signIn');
		} else {
			getLocation();
			getUserData();
		}
		return () => {
			getUserData();
		}
	}, [isFocused]);

	return (
		<SafeAreaView style={styles.container}>
			<ScrollView  contentContainerStyle={styles.headerView}
				refreshControl={
					<RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
				}>
				<TouchableOpacity onPress={() => {props.navigation.navigate('profile');}} style={styles.avatarView}>
					<Avatar size={50} image={proImage}/>
				</TouchableOpacity>
				<Image style={styles.image} source={{uri: `http://openweathermap.org/img/wn/${currentData?.weather[0]?.icon}@2x.png`}}/>
				<Text style={styles.statusText}>{currentData?.weather[0]?.description}</Text>
				{currentData && <Text style={styles.text}>{currentData?.name}, {currentData?.sys?.country}</Text>}
				<Text style={styles.tempText}>{(currentData?.main?.temp - 273.15).toFixed(0)}&deg;C</Text>
				<View style={styles.headerDetailsView}>
					<View style={[styles.headerData, {width: '30%'}]}>
						<Text style={styles.valueText}>{currentData?.main?.pressure}</Text>
						<Text style={styles.text}>Pressure</Text>
					</View>
					<View style={styles.hLine}/>
					<View style={styles.headerData}>
						<Text style={styles.valueText}>{currentData?.wind?.speed}</Text>
						<Text style={styles.text}>Wind Speed</Text>
					</View>
					<View style={styles.hLine}/>
					<View style={styles.headerData}>
						<Text style={styles.valueText}>{currentData?.main?.humidity}</Text>
						<Text style={styles.text}>Humidity</Text>
					</View>
				</View>
			</ScrollView>
			<ScrollView horizontal={true}>
				<BarChart
					style={styles.graph}
					data={dataChart}
					width={30 * 40 + 50}
					height={220}
					yAxisLabel=""
					chartConfig={chartConfig}
					bezier
					verticalLabelRotation={30}
				/>
			</ScrollView>
		</SafeAreaView>
	);

};

export default Main;
