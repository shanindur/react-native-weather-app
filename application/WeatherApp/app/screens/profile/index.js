/**
 * Weather App
 *
 * Profile
 *
 * @author       Shanindu Rajapaksha
 *
 * @version      0.1.0 2020-Jul-01
 */
import React, { useEffect, useState } from 'react';
import {
	SafeAreaView,
	ScrollView,
	View,
	Text,
	Modal,
	TouchableOpacity
} from 'react-native';
import {useIsFocused} from '@react-navigation/native';
import {Avatar, SquareButton, TextInput } from '../../components';
import ImagePicker from 'react-native-image-picker';
import styles from './styles';
import {StorageService} from '../../services';
import { Colors } from '../../util';
import { showMessage } from 'react-native-flash-message';
import Icon from 'react-native-vector-icons/FontAwesome5';
import { useDispatch } from 'react-redux';
import { SET_SIGN_OUT_STATE } from '../../redux/actions/actionType';

const options = {
	title: 'Select Image',
	storageOptions: {
		skipBackup: true,
		path: 'images'
	},
	quality: 1,
	maxWidth: 200,
	maxHeight: 200
};

const Profile = props => {
	const isFocused = useIsFocused();
	const [firstName, setFirstName] = useState();
	const [lastName, setLastName] = useState();
	const [mobile, setMobile] = useState();
	const [proImage, setProImage] = useState();
	const [editModalVisible, setEditModalVisible] = useState(false);

	const dispatch = useDispatch();

	const editProfilePicture = () => {
		ImagePicker.showImagePicker(options, (response) => {
			if (response.didCancel) {
				// console.log('User cancelled image picker');
			} else if (response.error) {
				// console.log('ImagePicker Error: ', response.error);
				showMessage({
					message: `${response.error}, Please go to settings and give permission to use device images.`,
					type: 'danger',
					duration: 5000
				});
			} else if (response.customButton) {
				// console.log('User tapped custom button: ', response.customButton);
			} else {
				setProImage(`data:image/webp;base64,${response.data}`);
				saveUser();
			}
		});
	};

	const saveUser = async() => {
		const USER = {
			fname: firstName,
			lname: lastName,
			mobile: mobile,
			profile: proImage
		};
		await StorageService.storeData('USER', USER);

		setEditModalVisible(false);
		showMessage({
			message: 'Your details was saved successfully!',
			type: 'success',
			duration: 5000
		});
		getUserData();

	};

	const getUserData = async() => {
		const user = await StorageService.getData('USER');
		if (user){
			setFirstName(user.fname);
			setLastName(user.lname);
			setMobile(user.mobile);
			setProImage(user.profile);
		}

	};

	useEffect(() => {
		getUserData();
		return () => {
			getUserData();
			saveUser();
		};
	}, [isFocused]);

	return (
		<SafeAreaView style={styles.container}>
			<View style={styles.headerView}>
				<TouchableOpacity onPress={() => props.navigation.goBack()} style={{width:'90%'}}>
					<Icon color={Colors.primaryColour} name={'arrow-left'} size={20} />
				</TouchableOpacity>
				<TouchableOpacity onPress={() => {props.navigation.replace('signIn'); dispatch({type: SET_SIGN_OUT_STATE});} }>
					<Text style={styles.signOut}>Logout</Text>
				</TouchableOpacity>
			</View>
			<Text style={styles.title}>Profile</Text>
			<ScrollView>
				<View style={styles.photoView}>
					<View style={styles.photoAvatar}>
						<Avatar isEditable={true} clickOnEditProfile={() => editProfilePicture()} size={150} image={proImage}/>
					</View>
				</View>
				<View style={styles.detailView}>
					<View style={styles.nameView}>
						<Text style={styles.text}>{firstName}</Text>
						<View style={styles.span}/>
						<Text style={styles.text}>{lastName}</Text>
					</View>
					<Text style={styles.mobileText}>{mobile}</Text>
				</View>
				<View style={styles.buttonView}>
					<SquareButton onPress={() => setEditModalVisible(true)} rounded={true} text={'Edit'} backgroundColor={Colors.primaryColor} buttonWidth={'30%'} fontColor={Colors.white} fontSize={18} />
				</View>
			</ScrollView>
			<Modal
				animationType={'fade'}
				transparent={true}
				visible={editModalVisible}
			>
				<View style={styles.overlay}>
					<TouchableOpacity
						style={styles.overlayTouchable}
						onPress={() => setEditModalVisible(false)}
					/>
				</View>
				<View style={styles.modalContentWrapper}>
					<View style={[styles.modalContent]}>
						<Text style={styles.title}>Edit User</Text>
					</View>
					<View style={styles.modalDetailContent}>
						<TextInput testID={'firstName'} placeholder={'First Name'} iconName={'user-o'} onChangeText={(value) => setFirstName(value)} value={firstName}/>
						<TextInput testID={'lastName'} placeholder={'Last Name'} iconName={'user-o'} onChangeText={(value) => setLastName(value)} value={lastName}/>
						<TextInput testID={'mobile'} placeholder={'Mobile'} iconName={'phone'} onChangeText={(value) => setMobile(value)} value={mobile}/>
					</View>
					<View style={styles.buttonView}>
						<SquareButton onPress={() => saveUser()} rounded={true} text={'Update'} backgroundColor={Colors.primaryColor} buttonWidth={'50%'} fontColor={Colors.white} fontSize={18} />
					</View>
				</View>
			</Modal>
		</SafeAreaView>
	);

};

export default Profile;
