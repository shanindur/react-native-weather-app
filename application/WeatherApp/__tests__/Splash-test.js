// __tests__/__snapshot__/Splash-test.js
import React from 'react';
import {SplashScreen} from '../app/screens';
import renderer from 'react-test-renderer';

global.fetch = jest.fn(() => new Promise(resolve => resolve()));
jest.mock('react-native-gesture-handler', () => {});
jest.mock('react-native-image-picker', () => {});
jest.mock('@react-native-community/async-storage', () => {});
jest.mock('@react-native-community/geolocation', () => {});

test('Snapshot test', () => {
	const tree = renderer.create(<SplashScreen />).toJSON();
	expect(tree).toMatchSnapshot();
});
