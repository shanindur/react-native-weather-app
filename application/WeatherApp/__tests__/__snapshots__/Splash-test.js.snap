// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../app/assets/images/splash.jpg",
        }
      }
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": "100%",
            "width": "100%",
          },
          undefined,
        ]
      }
    />
  </View>
</RCTSafeAreaView>
`;
